client/src/main/java/io/pravega/client/stream/impl/PositionImpl.java

```diff
@@ -55,37 +62,42 @@ public PositionImpl(Map<SegmentWithRange, Long> segments) {
     }
     
     @Builder(builderClassName = "PositionBuilder")
-    PositionImpl(Map<Segment, Long> ownedSegments, Map<Segment, Range> segmentRanges) {
-        this.ownedSegments = ownedSegments;
+    PositionImpl(Map<Segment, Long> ownedSegments, Map<Segment, Range> segmentRanges, List<Entry<Segment, Long>> updatesToSegmentOffsets) {
+        this.ownedSegments = Collections.unmodifiableMap(ownedSegments);
+        this.updatesToSegmentOffsets = Collections.unmodifiableList((updatesToSegmentOffsets == null) ? Collections.emptyList() : updatesToSegmentOffsets);
```

Comment:
This computation is wrong. The updatesToSegmentOffsets list is passed in without copying it. This means that if a caller holds onto one position object, and then calls readNext the first position object will be updated and have the wrong value. Please add a test for this.

Reply:
I have added an integration test to verify that the concurrent access and appends to `updatesToSegmentOffsets`. It shows that the optimization is safe and the offsets are correctly calculated.

Reply:
Perhaps I am missing something, but I don't see any actual concurrency in the test.

Reply:
Now the new test (`integration/ReadTest.java`) involves different threads for writing events, reading events and checking positions objects, so all this activity should happen concurrently. If there is any issue in `PositionImpl` when reading the `segmentOffsetUpdates` list is also being used by `EventStreamReaderImpl`, the test should fail.